packages/core/components/Modal/ModalProvider.tsx

```diff
@@ -1,26 +1,91 @@
+import { nanoid } from 'nanoid';
 import * as React from 'react';
 
 import ModalContext, { Windows } from './ModalContext';
 
 export type ModalProviderProps = {};
 
-const ModalProvider: React.FunctionComponent<ModalProviderProps> = ({
-  children,
-}) => {
-  const [windows, setWindows] = React.useState<Array<Windows>>([]);
+type WindowStack = Record<string, Windows>;
+
+type WindowAction =
+  | {
+      type: 'ADD_WINDOW';
+      id: string;
+      window: Windows;
+    }
+  | {
+      type: 'UPDATE_WINDOW';
+      id: string;
+      window: Windows;
+    }
+  | {
+      type: 'REMOVE_WINDOW';
+      id: string;
+    }
+  | {
+      type: 'REMOVE_BY_TITLE';
+      title: string;
+    };
+
+const windowStackReducer: React.Reducer<WindowStack, WindowAction> = (
+  state,
+  action,
+) => {
+  const newWindows = { ...state };
+  switch (action.type) {
+    case 'ADD_WINDOW': {
+      newWindows[action.id] = action.window;
+      return newWindows;
+    }
+    case 'REMOVE_WINDOW': {
+      delete newWindows[action.id];
+      return newWindows;
+    }
+    case 'UPDATE_WINDOW': {
+      newWindows[action.id] = action.window;
+      return newWindows;
+    }
+    case 'REMOVE_BY_TITLE': {
+      Object.entries(newWindows).forEach(([id, window]) => {
+        if (window.title === action.title) {
+          delete newWindows[id];
+        }
+      });
+      return newWindows;
```

Comment:
wouldn't `.filter` be better?

*_it is not a suggestion, btw_

Reply:
`.filter` doesn't really work here since `state`/`newWindows` is an object map rather than an array. It could maybe be done by filtering on the object keys, then reassigning the object keys. But that's not really a clear pattern to read 😕